### PR TITLE
Make `BruteForceSampler` stateless

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
Currently, `BruteForceSampler` uses `study.enqueue` mechanism to run a distributed breadth-first search. This causes problems when `BruteForceSampler` is used together with other samplers or with `enqueue_trial`/`add_trial`. Instead, we build a search-space tree every time `suggest_*` is called and sample an unexpanded node.

## Description of the changes
We add a `_TreeNode` class that explicitly handles the search-space tree, and sample an unexpanded node in `sample_independent`.

### CODE REVIEW

**Comment by contramundum53** [comment]
@not522 @toshihikoyanase Could you review this PR?
Feel free to unassign yourself if you are busy.

**Comment by contramundum53** [comment]
@cross32768 Could you review this PR?

**Comment by cross32768** [comment]
Sorry for the extreme example, if all parameters that are candidates for bruteforce are added through `add_trials`, it looks same parameter are proposed by this sampler. Is this expected behavior?

Below is example code.
```
import optuna
from optuna import samplers
from optuna.trial import Trial

def objective(trial: Trial) -> float:
    a = trial.suggest_int("a", 0, 1)
    return a

study = optuna.create_study(sampler=samplers.BruteForceSampler())

# Manually add a trial. This should not be tried again.
study.add_trial(
    optuna.create_trial(
        params={"a": 0},
        value=200.0,
        distributions={
            "a": optuna.distributions.IntDistribution(0, 1),
        },
    )
)

# Manually add a trial. This should not be tried again.
study.add_trial(
    optuna.create_trial(
        params={"a": 1},
        value=200.0,
        distributions={
            "a": optuna.distributions.IntDistribution(0, 1),
        },
    )
)

study.optimize(objective)
import pprint
pprint.pprint(study.trials)
```

This code shows result as shown below.
```
[FrozenTrial(number=0, state=TrialState.COMPLETE, values=[200.0], datetime_start=datetime.datetime(2023, 2, 20, 16, 37, 56, 543513), datetime_complete=datetime.datetime(2023, 2, 20, 16, 37, 56, 543513), params={'a': 0}, user_attrs={}, system_attrs={}, intermediate_values={}, distributions={'a': IntDistribution(high=1, log=False, low=0, step=1)}, trial_id=0, value=None),
 FrozenTrial(number=1, state=TrialState.COMPLETE, values=[200.0], datetime_start=datetime.datetime(2023, 2, 20, 16, 37, 56, 543751), datetime_complete=datetime.datetime(2023, 2, 20, 16, 37, 56, 543751), params={'a': 1}, user_attrs={}, system_attrs={}, intermediate_values={}, distributions={'a': IntDistribution(high=1, log=False, low=0, step=1)}, trial_id=1, value=None),
 FrozenTrial(number=2, state=TrialState.COMPLETE, values=[1.0], datetime_start=datetime.datetime(2023, 2, 20, 16, 37, 56, 543925), datetime_complete=datetime.datetime(2023, 2, 20, 16, 37, 56, 545389), params={'a': 1}, user_attrs={}, system_attrs={}, intermediate_values={}, distributions={'a': IntDistribution(high=1, log=False, low=0, step=1)}, trial_id=2, value=None)]
```

**Comment by cross32768** [comment]
I found that depending trials added by user, study.optimize() continues eternally. It is difficult to explain, see the code below.

Normal Pattern:

```
import optuna
from optuna import samplers
from optuna.trial import Trial

def objective(trial: Trial) -> float:
    a = trial.suggest_int("a", 0, 1)

    if a == 0:
        b = trial.suggest_float("b", -1.0, 1.0, step=0.5)
        return a + b
    else:
        return a

study = optuna.create_study(sampler=samplers.BruteForceSampler(seed=6))

# Manually add a trial. This should not be tried again.
study.add_trial(
    optuna.create_trial(
        params={"a": 0, "b": -1.0},
        value=0.0,
        distributions={
            "a": optuna.distributions.IntDistribution(0, 1),
            "b": optuna.distributions.FloatDistribution(-1.0, 1.0, step=0.5),
        },
    )
)

study.optimize(objective)
```

This code outputs below.
```
/tmp/ipykernel_7915/566721632.py:14: ExperimentalWarning: BruteForceSampler is experimental (supported from v3.1.0). The interface can change in the future.
  study = optuna.create_study(sampler=samplers.BruteForceSampler(seed=6))
[I 2023-02-20 16:56:20,648] A new study created in memory with name: no-name-cefdee1c-dca5-4575-bddc-2540f38bea09
[I 2023-02-20 16:56:20,654] Trial 1 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 0 with value: 0.0.
[I 2023-02-20 16:56:20,656] Trial 2 finished with value: 1.0 and parameters: {'a': 0, 'b': 1.0}. Best is trial 0 with value: 0.0.
[I 2023-02-20 16:56:20,657] Trial 3 finished with value: -0.5 and parameters: {'a': 0, 'b': -0.5}. Best is trial 3 with value: -0.5.
[I 2023-02-20 16:56:20,658] Trial 4 finished with value: 0.5 and parameters: {'a': 0, 'b': 0.5}. Best is trial 3 with value: -0.5.
[I 2023-02-20 16:56:20,660] Trial 5 finished with value: 0.0 and parameters: {'a': 0, 'b': 0.0}. Best is trial 3 with value: -0.5.
```

Abnormal? Pattern.

```
import optuna
from optuna import samplers
from optuna.trial import Trial

def objective(trial: Trial) -> float:
    a = trial.suggest_int("a", 0, 1)

    if a == 0:
        b = trial.suggest_float("b", -1.0, 1.0, step=0.5)
        return a + b
    else:
        return a

study = optuna.create_study(sampler=samplers.BruteForceSampler(seed=6))

# Manually add a trial. This should not be tried again.
study.add_trial(
    optuna.create_trial(
        params={"a": 1, "b": -1.0},  # CHANGE "a": 0 → "a": 1
        value=0.0,
        distributions={
            "a": optuna.distributions.IntDistribution(0, 1),
            "b": optuna.distributions.FloatDistribution(-1.0, 1.0, step=0.5),
        },
    )
)

study.optimize(objective)
```

This code outputs below.

```
/tmp/ipykernel_7915/3908994821.py:14: ExperimentalWarning: BruteForceSampler is experimental (supported from v3.1.0). The interface can change in the future.
  study = optuna.create_study(sampler=samplers.BruteForceSampler(seed=6))
[I 2023-02-20 16:57:11,516] A new study created in memory with name: no-name-c853bec2-42fe-4022-b9d0-03f4450666da
[I 2023-02-20 16:57:11,517] Trial 1 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 0 with value: 0.0.
[I 2023-02-20 16:57:11,519] Trial 2 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 0 with value: 0.0.
[I 2023-02-20 16:57:11,520] Trial 3 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 0 with value: 0.0.
[I 2023-02-20 16:57:11,522] Trial 4 finished with value: -1.0 and parameters: {'a': 0, 'b': -1.0}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,523] Trial 5 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,525] Trial 6 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,526] Trial 7 finished with value: 0.0 and parameters: {'a': 0, 'b': 0.0}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,527] Trial 8 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,529] Trial 9 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,530] Trial 10 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,532] Trial 11 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,533] Trial 12 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,534] Trial 13 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,536] Trial 14 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,537] Trial 15 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,539] Trial 16 finished with value: 1.0 and parameters: {'a': 0, 'b': 1.0}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,540] Trial 17 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,542] Trial 18 finished with value: 0.5 and parameters: {'a': 0, 'b': 0.5}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,544] Trial 19 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,545] Trial 20 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,546] Trial 21 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,548] Trial 22 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,550] Trial 23 finished with value: -0.5 and parameters: {'a': 0, 'b': -0.5}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,551] Trial 24 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,553] Trial 25 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,554] Trial 26 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,556] Trial 27 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,557] Trial 28 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,559] Trial 29 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,561] Trial 30 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
[I 2023-02-20 16:57:11,562] Trial 31 finished with value: 1.0 and parameters: {'a': 1}. Best is trial 4 with value: -1.0.
...
```

**Comment by contramundum53** [comment]
More detailed explanation: Currently BruteForceSampler does not make distinction between the trials that it suggested and the trials that are enqueued manually. Also, BruteForceSampler does not consider the orders of trials. Now think the situation that we have instead enqueued N `{"a": 1}` trials and BruteForceSampler suggested one `{"a": 1, "b": 1}` trial. Shouldn't we continue searching in this case, because we need to know the values of `{"a": 1, "b": 0}` and `{"a": 1, "b": -1}`?

**Comment by contramundum53** [comment]
I think that a stateless brute force sampler should not make distinction between this case and your case.

**Comment by contramundum53** [comment]
Or should we raise an error when such trial is found?

**Comment by cross32768** [comment]
Yes, I think to raise error is better when such a trial is wrongly inserted.

**Comment by contramundum53** [comment]
Thank you for your comment!
I fixed it to raise an error. PTAL.

File: optuna/samplers/_brute_force.py [hunk 1]
```diff
@@ -179,6 +181,7 @@ def sample_independent(
         param_name: str,
         param_distribution: BaseDistribution,
     ) -> Any:
+        print("hoge")
```
Comment (cross32768) [inline optuna/samplers/_brute_force.py:184]: Is this remaining debug message?

File: optuna/samplers/_brute_force.py [hunk 2]
```diff
@@ -189,6 +192,7 @@ def sample_independent(
         tree = self._build_tree((t for t in trials if t.number != trial.number), trial.params)
         candidates = _enumerate_candidates(param_distribution)
         tree.expand(param_name, candidates)
+        print(tree)
```
Comment (cross32768) [inline optuna/samplers/_brute_force.py:195]: ditto

**Comment by cross32768** [comment]
It looks that still current code does not raise error for the case shown below (the case which study.optimize() continues eternally).

```
import optuna
from optuna import samplers
from optuna.trial import Trial

def objective(trial: Trial) -> float:
    a = trial.suggest_int("a", 0, 1)

    if a == 0:
        b = trial.suggest_float("b", -1.0, 1.0, step=0.5)
        return a + b
    else:
        return a

study = optuna.create_study(sampler=samplers.BruteForceSampler(seed=6))

# Manually add a trial. This should not be tried again.
study.add_trial(
    optuna.create_trial(
        params={"a": 1, "b": -1.0},  # CHANGE "a": 0 → "a": 1
        value=0.0,
        distributions={
            "a": optuna.distributions.IntDistribution(0, 1),
            "b": optuna.distributions.FloatDistribution(-1.0, 1.0, step=0.5),
        },
    )
)

study.optimize(objective)
```

**Comment by cross32768** [comment]
I confirmed that the new commit raises error for the abnormal pattern in https://github.com/optuna/optuna/pull/4408#issuecomment-1436502517 .

File: optuna/samplers/_brute_force.py [hunk 3]
```diff
@@ -49,11 +118,8 @@ def objective(trial):
         :func:`~optuna.trial.Trial.suggest_float`, ``step=None`` is not allowed.
 
     Note:
-        This sampler assumes that it suggests all parameters and that the search space is fixed.
-        For example, the sampler may fail to try the entire search space in the following cases.
-
-        * Using with other samplers or :meth:`~optuna.study.Study.enqueue_trial`
-        * Changing suggestion ranges or adding parameters in the same :class:`~optuna.study.Study`
+        The sampler may fail to try the entire search space in when the suggestion ranges or
+        parameters are changed in the same :class:`~optuna.study.Study`
```
Comment (contramundum53) [inline optuna/samplers/_brute_force.py:122]:
```suggestion
        parameters are changed in the same :class:`~optuna.study.Study`.
```

File: optuna/samplers/_brute_force.py [hunk 4]
```diff
@@ -76,24 +142,61 @@ def sample_relative(
...
+        return tree
+
     def sample_independent(
         self,
         study: Study,
         trial: FrozenTrial,
         param_name: str,
         param_distribution: BaseDistribution,
     ) -> Any:
+        print("hoge")
```
Comment (contramundum53) [inline optuna/samplers/_brute_force.py:184]:
```suggestion
```

File: optuna/samplers/_brute_force.py [hunk 5]
```diff
@@ -76,24 +142,61 @@ def sample_relative(
...
-        self._rng.shuffle(candidates)
-
-        for value in candidates[1:]:
-            params = trial.params.copy()
-            params[param_name] = value
-            study.enqueue_trial(params, skip_if_exists=True)
-
-        return candidates[0]
+        tree.expand(param_name, candidates)
+        print(tree)
```
Comment (contramundum53) [inline optuna/samplers/_brute_force.py:195]:
```suggestion
```

File: optuna/samplers/_brute_force.py (hunk 4 above)
Comment (contramundum53) [reply optuna/samplers/_brute_force.py]: Thank you for pointing out!

**Review by not522 — APPROVED** [review]
LGTM

**Pull request merged**